### PR TITLE
quin - time

### PR DESCRIPTION
**Pull request by quinqu**
# Assignment Submission: Hotel

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | pretty much all of it, i had a lot of trouble implementing room block|
What was a design decision you made that changed over time over the project? | I wanted more stuff in the room and reservation class but most of the functions remained in the hotel class | 
What was a concept you gained clarity on, or a learning that you'd like to share? | I really figured out that more design planning is needed for OOP | 
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | I stored reservations for room in an array in the room object, I tested the length of reservations array of a room once. I also checked if an object was an the object intended. a nominal case is when | 
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | An edge case test I created was making sure my code could start a reservation when the same one ends. When there is an extreme operating parameter is when there is an edge case (cite:  stackexchange.com)  | 
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I think I did terrible at the pseudocode and just jumped in. Writing tests was the easiest part of this assignment. I don't think my project structure was very great and I should have actually redesigned it, but it was too late.   |

### CODE REVIEW

File: lib/reservation.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    end 
+
+    def get_total_cost 
+      return @total_cost
+    end 
+
+    def check_dates(start_date, end_date)
+      if Date.today > start_date || Date.today > end_date
+        raise ArgumentError.new("Invalid date, you can't start or end a reservation in the past")
+      end
```
Comment (kaidamasaki) [inline lib/reservation.rb:23]: If you're going to have a check like this please make sure that your example dates in your tests are in the future.  (Either by placing them far in the future or making them relative to `Date.today`.)  

Several of your tests fail because dates are now in the past.

File: test/hotel_test.rb [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+require_relative 'test_helper'
+
+
+
+Minitest::Reporters.use! Minitest::Reporters::SpecReporter.new
+
+describe "initialize" do 
+  it "I can access the list of all of the rooms in the hotel" do 
+    hotel = HotelSystem::Hotel.new 
+    hotel.must_be_instance_of HotelSystem::Hotel
```
Comment (kaidamasaki) [inline test/hotel_test.rb:13]: Please use `expect` when writing assertions:

```suggestion
    expect(hotel).must_be_instance_of HotelSystem::Hotel
```

Minitest generates lots of warnings otherwise.  (And in future versions of Minitest this won't work.)

File: lib/date_range.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+    def nights
+      return 3
+    end
+
+  end 
+
+
+
+end
```
Comment (kaidamasaki) [inline lib/date_range.rb:26]: This looks like it's left over from the scaffolding.

In the future if you're not using code you should remove it.

File: lib/hotel.rb [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+      end  
+      return list_reservations
+    end 
+    
+    def find_room(room_number)
+      @rooms.each do |room| 
+        if room_number == room.room_number 
+          return room
+        end 
+      end
```
Comment (kaidamasaki) [inline lib/hotel.rb:58]: You can simplify this using `find`:

```suggestion
      @rooms.find do |room| 
        room_number == room.room_number
      end
```

File: lib/hotel.rb [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    end
+
+    def find_reservation(reservation_number)
+      @rooms.each do |room|
+        room.reservations.each do |reservation|
+          if reservation.reservation_number == reservation_number
+            return reservation
+          end 
+        end 
+      end
```
Comment (kaidamasaki) [inline lib/hotel.rb:68]: If you're going to have `reservations` stored in `Room` you should have this functionality inside of a method there instead of here:

```suggestion
      @rooms.each do |room|
        room.find_reservation(reservation_number)
      end
```

That properly encapsulates this functionality and means that `Hotel` has to know less about `Room`.

File: lib/hotel.rb [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+      if true_for_all.length == room.reservations.length 
+        room.reservations << HotelSystem::Reservation.new(next_reservation_number, start_date, end_date)
+      else 
+        raise ArgumentError.new("you can't overlap reservations")
+      end 
+    end
+
+    
+  #there may be some cases where a reservation starts when one ends, but trust me it still works the method name isn't very accurate
+    def no_shared_days?(start_date, end_date, reservation)
```
Comment (kaidamasaki) [inline lib/hotel.rb:112]: Putting this sort of functionality inside of `DateRange` could have helped with cleaning up this code and with following the Single Responsibility Principle more closely.

File: lib/hotel.rb [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+      return room.cost_per_night - (room.cost_per_night*(rate/100))
+    end 
+  end 
+end 
+
+
+def main
+end
+
+main
```
Comment (kaidamasaki) [inline lib/hotel.rb:163]: Please remove method stubs like this in the future.

File: lib/hotel.rb [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+          reservation_start = reservation_start +  1
+       end 
+       reservation_start = reservation.start_date 
+       start_date += 1
+      end 
+      if (reservation.end_date - reservation.start_date == 1) && reservation_start == start_date
+        return false 
+      else
+        return count < 2
+      end
```
Comment (kaidamasaki) [inline lib/hotel.rb:130]: This logic is very complex.  Please comment the individual parts of something like this in the future.

File: lib/hotel.rb [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+        return count < 2
+      end 
+    end 
+     
+    private
+
+    #used for fine_reservations_for_room and list_reservations_with_date 
+    def is_between_two_dates(range_wanted, date)
+      return date >= range_wanted[0] && date <= range_wanted[1];
+    end
```
Comment (kaidamasaki) [inline lib/hotel.rb:138]: This method would have been cleaner if it compared two `DateRange` objects.  (Instead of a two element array of `Date` objects with a single date.)

File: lib/hotel.rb [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+      @rooms.each do |room|
+        room.reservations.each do |reservation| 
+          if max < reservation.reservation_number
+            max = reservation.reservation_number
+          end 
+
+        end
+      end
+      return max + 1 
+    end
```
Comment (kaidamasaki) [inline lib/hotel.rb:151]: This can be simplified using `max`:

```suggestion
    def next_reservation_number  
      max = @rooms.max do |room|
        room.reservations.max do |reservation| 
          reservation.reservation_number
        end
      end
      return max + 1 
    end
```

File: lib/reservation.rb [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+  class Reservation 
+    attr_reader :start_date, :end_date, :total_cost 
+    attr_accessor :reservation_number
+
+    def initialize(reservation_number, start_date, end_date)
+      @reservation_number = reservation_number
+      check_dates(start_date, end_date)
+      @start_date = start_date
+      @end_date = end_date
+      @total_cost = ((@end_date - @start_date)) * 200
```
Comment (kaidamasaki) [inline lib/reservation.rb:13]: Extra parens:

```suggestion
      @total_cost = (@end_date - @start_date) * 200
```

File: test/hotel_test.rb [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+
+  end 
+
+  it "finds an array of reservations" do 
+  
+    hotel = HotelSystem::Hotel.new
+
+    room = hotel.find_room(1)
+    
+    res1 = HotelSystem::Reservation.new(1, Date.today, Date.new(2020,03,29))
```
Comment (kaidamasaki) [inline test/hotel_test.rb:40]: This test will break after March 29th 2020:

```suggestion
    res1 = HotelSystem::Reservation.new(1, Date.today, Date.today + 15)
```

**Comment by kaidamasaki** [comment]
# Hotel






## Section 1: Major Learning Goals



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
Practices SRP by having at least two separate classes with distinct responsibilities, and test files for these two classes | ✔️
Overall, demonstrates understanding instance variables vs. local variables. (There aren't unnecessarily too many instance variables, when it should be a local variable) | ✔️
For each test file, tests demonstrate an understanding of instantiating objects properly, and using Arrange-Act-Assert | ✔️
Practices pseudocode and TDD, and reflected on it by filling out the reflection questions | Didn't pseudocode.  Please try to practice this in the future.  It's easier to redesign psuedocode than a full implementation.
Practices git with at least 15 small commits and meaningful commit messages | Going forward please commit more often and leave more detailed messages.  This will help you going forward, especially in group projects.

## Section 2: Code Review and Testing Requirements

| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
There is a class that represents a reservation, and a second class that holds/manages a collection of reservations through composition (instance variable) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date is complex logic that is separated into method(s) (and potentially class(es)) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date has unit tests | ✔️
All of the given tests run and pass | ✔️ Tests only pass if I set my clock back, but they do all pass on March 9th.
A test coverage tool is installed and used, and shows 95% test coverage | ✔️ Tests only pass if I set my clock back, but they do all pass on March 9th.

## Section 3: Feature Requirements

| Feature Requirement: There is a method that... | yes/no |
| --- | --- |
gives back a list of rooms, and it's tested | No tests.
creates a specific reservation for a room for a given date range, and it has nominal test cases | ✔️
creates a specific reservation for a room for a given date range, and it tests an edge case, such as no available room, or invalid date range | ✔️
gives back a list of reservations on a given date. Its tests include a test that makes several reservations for a given date | Not enough edge cases tested.
calculates the total price for a reservation | ✔️
gives back a list of available rooms for a given date range, and it has nominal test cases | ✔️
gives back a list of available rooms for a given date range, and it has edge test cases, such as no available room, or invalid date range | No edge case test.
creates a block of rooms | Not implemented.
reserves a room from a block | Not implemented.

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ total in all sections |
| Yellow (Approaches Standards) | 9-13 total in all sections | ✔️
| Red (Not at Standard) | 0-8 total in all sections, or assignment is breaking/doesn’t run with less than 5 minutes of debugging |

### Additional Feedback



Great work overall! You've built your first project with minimal starting code. This represents an incredible milestone in your journey, and you should be proud of yourself!

I am particularly impressed by the way that you reflected upon your code and the quality of the tests that do exist.

I do see some room for improvement around testing edge cases and making sure to take a step back and refactor as code starts to get unwieldy.

## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever |
| Descriptive/Readable | ✅
| Concise |
| Logical/Organized |

**Review by kaidamasaki — COMMENTED** [review]
Good job.  Some parts of your design were a little messy but it seems like you're aware of that based on your refactors.txt.  

Here are some tips for how to clean up your code.
